src/main/java/com/cleanroommc/groovyscript/compat/mods/prodigytech/ExplosionFurnace.java

```diff
@@ -0,0 +1,121 @@
+    }
+
+    public void addRecipe(ExplosionFurnaceManager.ExplosionFurnaceRecipe x) {
+        addScripted(x);
+        ExplosionFurnaceManager.addRecipe(x);
+    }
+
+    @MethodDescription(example = @Example("item('prodigytech:ferramic_ingot')"))
+    public void removeByOutput(ItemStack output) {
+        ExplosionFurnaceManager.RECIPES.removeIf(r -> {
+            if (!r.getOutput().isItemEqual(output)) return false;
+            addBackup(r);
+            return true;
+        });
+    }
+
+    @MethodDescription(example = @Example(priority = 2000, commented = true))
+    public void removeAll() {
+        ExplosionFurnaceManager.RECIPES.forEach(this::addBackup);
+        ExplosionFurnaceManager.removeAllRecipes();
+    }
+
+    @MethodDescription(type = MethodDescription.Type.QUERY)
+    public SimpleObjectStream<ExplosionFurnaceManager.ExplosionFurnaceRecipe> streamRecipes() {
+        return new SimpleObjectStream<>(ExplosionFurnaceManager.RECIPES)
+                .setRemover(this::remove);
+    }
+
+    @Property(property = "input", valid = {@Comp(type = Comp.Type.GTE, value = "1"), @Comp(type = Comp.Type.LTE, value = "2")})
+    @Property(property = "output", valid = @Comp("1"))
+    public static class RecipeBuilder extends AbstractRecipeBuilder<ExplosionFurnaceManager.ExplosionFurnaceRecipe> {
+
+        @Property(valid = @Comp(value = "0", type = Comp.Type.GTE))
+        private int craftPerReagent = 0;
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE))
+        private int power;
+
+        @RecipeBuilderMethodDescription
+        public ExplosionFurnace.RecipeBuilder craftPerReagent(int craftPerReagent) {
+            this.craftPerReagent = craftPerReagent;
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public ExplosionFurnace.RecipeBuilder power(int power) {
+            this.power = power;
+            return this;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return "Error adding ProdigyTech Explosion Furnace Recipe";
+        }
+
+        @Override
+        public void validate(GroovyLog.Msg msg) {
+            validateItems(msg, 1, 2, 1, 1);
+            validateFluids(msg);
+            msg.add(craftPerReagent > 0 && input.size() == 1, "craftPerReagent is only allowed with 2 inputs!");
+            if (input.size() == 2 && craftPerReagent <= 0) {
+                craftPerReagent = 1;
+            }
+        }
+
+        @Override
+        @RecipeBuilderRegistrationMethod
+        public @Nullable ExplosionFurnaceManager.ExplosionFurnaceRecipe register() {
+            if (!validate()) return null;
+            ExplosionFurnaceManager.ExplosionFurnaceRecipe recipe = null;
+            IIngredient inputItem = input.get(0);
+            // We do not do the OreDict check like in other places as it adds far too much code bloat
+            if (craftPerReagent == 0) {
```

Comment:
i think you should change this logic to be based on if `input.size() == 2`, and then change the defaultValue of `craftPerReagent` to `1`. this would allow you to simplify some of the logic used.